[WPE] WPE Platform: add support for gamepads

carlosgcampos opened the pull request:
#### 5d0654c0f1360fe46c85069cb964a19aaf846f1a
<pre>
[WPE] WPE Platform: add support for gamepads
<a href="https://bugs.webkit.org/show_bug.cgi?id=265647">https://bugs.webkit.org/show_bug.cgi?id=265647</a>

Reviewed by Adrian Perez de Castro.

Add WPEGamepadManager and WPEGamepad classes with a default
implementation using libmanette when available. Gamepads are now always
enabled by default in WPE and the required libwpe version for the old
API is checked with ifdefs when needed. This patch also adds the
GamepadProvider and PlatformGamepad implementations for WPE using the
new WPE platform API. We also had to make public the function to list
all WPEToplevels since it&apos;s required to find the WebPageProxy used for
gamepads.

Canonical link: <a href="https://commits.webkit.org/295094@main">https://commits.webkit.org/295094@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a114d58eaef19c8dd99d7bdaf66ad37e3dcf0edc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104095 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23799 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14119 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109291 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54763 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106135 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24167 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32343 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/79082 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107101 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18773 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93902 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59409 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18579 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11945 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54123 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/96770 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/12003 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111677 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/102706 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31251 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23042 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/88091 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31615 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90094 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87749 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32632 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10380 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25708 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16895 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31180 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/36492 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/126339 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30974 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34938 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34310 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32534 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alexgcastro commented on Source/WebKit/UIProcess/Gamepad/wpe/GamepadProviderWPE.cpp:
```diff
@@ -0,0 +1,181 @@
+        setShouldMakeGamepadsVisibile();
+
+    if (!m_inputNotificationTimer.isActive())
+        m_inputNotificationTimer.startOneShot(s_inputNotificationDelay);
```
Why do we need a 1ms delay?

carlosgcampos replied:
I'm not sure to be honest, it's what all other implementations do.

carlosgcampos replied:
Ah! I remember now, I think it's because sometimes the gamepads produce too many events, so at least with this we ensure we only report one event per millisecond

alexgcastro commented on Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
```diff
@@ -208,6 +211,33 @@ void ViewPlatform::requestExitFullScreen()
+            continue;
+
+        WPEView* view = nullptr;
+        wpe_toplevel_foreach_view(toplevel, +[](WPEToplevel* toplevel, WPEView* view, gpointer userData) -> gboolean {
```
Would it make sense to comment about the rationale of how we pick the view? I guess it is the one visible and focused? I can imagine situations where you can play and write messages in a different view but I don't think we should support that.

carlosgcampos replied:
Yes, we need to find the active window and then get the visible view that gets the input focus

alexgcastro reviewed (commented):
This is a great addition, some initial comments.

aperezdc commented on Source/WebKit/UIProcess/Gamepad/libwpe/UIGamepadProviderLibWPE.cpp:
```diff
@@ -27,28 +27,53 @@
 
-namespace WebKit {
+#if ENABLE(WPE_PLATFORM)
+static inline bool usingWPEPlatformAPI()
+{
+    return !!g_type_class_peek(WPE_TYPE_DISPLAY);
+}
```
We have a number of variables and similar functions sprinkled around the code (at least I could easily find instances in `WebPasteboardProxyWPE.cpp`, `DisplayVBlankMonitorDRM.cpp`, `WebProcessPoolGLib.cpp`, and `WebKitProtocolHandler.cpp`). While I do not think that is a blocker, we may want to factor them out into a single, shared implementation in the future.

carlosgcampos replied:
Yes, I thought about that, I'll find a common place to move this.

aperezdc commented on Source/WebKit/UIProcess/Gamepad/wpe/GamepadProviderWPE.cpp:
```diff
@@ -0,0 +1,181 @@
+namespace WebKit {
+using namespace WebCore;
+
+static const Seconds s_inputNotificationDelay { 1_ms };
```
This could be `constexpr`:

```suggestion
static constexpr Seconds s_inputNotificationDelay { 1_ms };
```

aperezdc commented on Source/WebKit/UIProcess/Gamepad/wpe/GamepadProviderWPE.cpp:
```diff
@@ -0,0 +1,181 @@
+
+void GamepadProviderWPE::playEffect(unsigned, const String&, GamepadHapticEffectType, const GamepadEffectParameters&, CompletionHandler<void(bool)>&& completionHandler)
+{
+    // Not supported by this provider.
```
I think it would make sense to add a `notImplemented()` marker here, because we could use [manette_device_rumble()](https://gnome.pages.gitlab.gnome.org/libmanette/doc/0.2/method.Device.rumble.html) in the future to complete the implementation. 

```suggestion
    // Not supported by this provider.
    notImplemented();
```

aperezdc reviewed (approved):
Patch LGTM, with a couple of very minor nits 🎮

carlosgcampos committed:
[WPE] WPE Platform: add support for gamepads
https://bugs.webkit.org/show_bug.cgi?id=265647

Reviewed by Adrian Perez de Castro.

Add WPEGamepadManager and WPEGamepad classes with a default
implementation using libmanette when available. Gamepads are now always
enabled by default in WPE and the required libwpe version for the old
API is checked with ifdefs when needed. This patch also adds the
GamepadProvider and PlatformGamepad implementations for WPE using the
new WPE platform API. We also had to make public the function to list
all WPEToplevels since it's required to find the WebPageProxy used for
gamepads.

Canonical link: https://commits.webkit.org/295094@main
